plugins/modules/cloud/scaleway/scaleway_function_namespace.py

```diff
@@ -0,0 +1,286 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+#
+# Scaleway Serverless function namespace management module
+#
+# Copyright (c) 2022, Guillaume MARTINEZ <lunik@tiwabbit.fr>
+# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
+module: scaleway_function_namespace
+short_description: Scaleway Function namespace management module
+version_added: 5.8.0
+author: Guillaume MARTINEZ (@Lunik)
+description:
+  - This module manages function namespaces on Scaleway account.
+extends_documentation_fragment:
+  - community.general.scaleway
+  - community.general.scaleway_waitable_resource
+
+
+options:
+  state:
+    type: str
+    description:
+      - Indicate desired state of the function namespace.
+    default: present
+    choices:
+      - present
+      - absent
+
+  project_id:
+    type: str
+    description:
+      - Project identifier.
+    required: true
+
+  region:
+    type: str
+    description:
+      - Scaleway region to use (for example C(fr-par)).
+    required: true
+    choices:
+      - fr-par
+      - nl-ams
+      - pl-waw
+
+  name:
+    type: str
+    description:
+      - Name of the function namespace.
+    required: true
+
+  description:
+    description:
+      - Description of the function namespace.
+    type: str
+
+  environment_variables:
+    description:
+      - Environment variables of the function namespace.
+      - Injected in functions at runtime.
+    type: dict
+
+  secret_environment_variables:
+    description:
+      - Secret environment variables of the function namespace.
+      - Updating thoses values will not output a C(changed) state in Ansible.
```

Comment:
Does that mean the values will be updated, but the module will not report that a change happened, or does it mean that the values will not be updated?

Reply:
See also https://github.com/ansible-collections/community.general/pull/5416#discussion_r1002756986.

Reply:
With your other comment I see that this sentence could be misleading.

With the actual algorithm :
- If there is another field that require an update, this field will be updated. The module will output as `changed`
- If only this field needs an update. We are still pushing an update to the API but report no change because we have no way to tell if the value really needed an update. The module will output `not changed`

Reply:
Are you sure about "If all fields doesn't need update but this one has a value specified we are still pushing an update to the API"? The way I understood the code yesterday that isn't the case. (Don't have time to look at it right now, but can take another look this evening.)

Reply:
If we leave `secret_environment_variable` on the side, the module should behave like a normal module : only POST update when needed.

When `secret_environment_variable` is involved, since we can't compare the value returned by the API (because the value is encrypted) to the value specified by the user (plain text), we have here a tricky situation.

I'm not totally satisfied with the current behavior of the module. I'm open to suggestions.

Reply:
It seems that the value returned by the API is hashed with `Argon2`. With the help of the package [argon2-cffi](https://pypi.org/project/argon2-cffi/) and a small verify function we should be able to verify if the value is changed or not.

For example : 
```python
from argon2 import PasswordHasher

ph = PasswordHasher()

hash="$argon2id$v=19$m=65536,t=1,p=2$GbzoEgbpN3Xv6/RbjOXDKQ$xc3FZ37FrxVV3dObjodVS10Fz6I9ctkCiPZWaOHZ60s"

assert ph.verify(hash, "value") # True if the value is match the hash
```

Reply:
To give more context.
Here is the value we POST to the API : 
```json
"secret_environment_variables": [
  {
    "key": "MY_SECRET_VAR",
    "value": "my_secret_value"
  }
]
```

And here is the value returned by the API :
```json
"secret_environment_variables": [
  {
    "key": "MY_SECRET_VAR",
    "hashed_value": "$argon2id$v=19$m=65536,t=1,p=2$BsikPnuMeoh5X2rLZD4y3g$b7o2a8b+aM62I6/IekNwtIjhCKJUEucHLiXB4dAcduc"
  }
]
```

Reply:
> I'm not totally satisfied with the current behavior of the module. I'm open to suggestions.

Hmm, I understand that. Implementing a password check to be able to do itempotency checks is probably the best solution though.

But you might want to try to use `passlib` instead of `argon2-cffi`, which supports Argon2 (https://passlib.readthedocs.io/en/stable/lib/passlib.hash.argon2.html), but also other hashes, so that in case the API eventually decides to switch to another hash you don't have to update the modules to be able to handle that.

Reply:
Understood 👍
Where do I need to specify the library requirements ?
Only in the module documentation or we should ship it as a requirement for the collection ?
Is `passlib` already in Ansible requirement ?

Reply:
This is a module requirement. The collection has no requirements declared (otherwise it would require hundreds of Python modules). I don't think passlib is a requirement of ansible-core, but even if it would be it wouldn't help since this is a module, not a plugin that's always executed on the contoller.

Reply:
I have implemented the hash verification in : 6effe81e6d1be87392e9544fd193a17706acc135
And more : 
- Some unit tests to validate added utils functions
- Required missing python lib checks